control/grpc.py

```
@@ -59,39 +59,23 @@
 
 
 class SubsystemsCache:
-    SUBSYSTEMS_CACHE_EXPIRATION = 30
-
-    def __init__(self, expiration=None):
+    def __init__(self):
         self.cache_lock = threading.Lock()
-        self.last_value_time = 0
-        self.subsystems = None
-        self.expiration = SubsystemsCache.SUBSYSTEMS_CACHE_EXPIRATION
-        if expiration is not None:
-            self.expiration = expiration
-
-    def _check_conditions(self) -> bool:
-        if not self.subsystems:
-            return False
-        if not self.last_value_time:
-            return False
-        if self.expiration and (time.time() - self.last_value_time >= self.expiration):
-            return False
-        return True
+        with self.cache_lock:
+            self.subsystems_info = None
 
     def get_subsystems(self):
         with self.cache_lock:
-            if not self._check_conditions():
-                return None
-            return self.subsystems
+            return self.subsystems_info
 
     def get_one_subsystem(self, subsys):
         if not subsys:
             return None
 
         with self.cache_lock:
-            if not self._check_conditions():
+            if not self.subsystems_info:
                 return None
-            for s in self.subsystems:
+            for s in self.subsystems_info.subsystems:
                 try:
                     if s["nqn"] == subsys:
```

SubsystemsCache.set_subsystems stores pb2.subsystems_info objects; each entry in self.subsystems_info.subsystems is a protobuf message, not a dict. Yet get_one_subsystem iterates over those messages and tries to subscript them (s["nqn"]) inside a bare try/except. That raises TypeError every time, the exception is swallowed, and the loop falls through to return None. Downstream callers such as list_hosts_safe (control/grpc.py, around L5024) and list_connection_for_one_subsystem (~L5153) always get None, so they re-issue the  SPDK RPC instead of using the cache.

---

Nice catch. I changed the type of the cached array and forgot to change it over there as well.

---

I don't know where you're looking but there is no s["nqn"]any more, it's s.nqn.